feat: add `lapack/base/dlasv2`

Pranavchiku opened the pull request:
Towards #2464.

## Description

> What is the purpose of this pull request?

This pull request adds JS implementation for `lapack/base/dlasv2`

## Related Issues

> Does this pull request have any related issues?

NA

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Pranavchiku committed:
feat: init base implementation

Pranavchiku committed:
feat: init public apis in lib

Pranavchiku committed:
docs: add README

Pranavchiku committed:
docs: add repl

Pranavchiku committed:
feat: add other files

Pranavchiku committed:
chore: add type notation in jsdoc of ndarray

Pranavchiku commented on lib/node_modules/@stdlib/lapack/base/dlasv2/lib/base.js:
```diff
@@ -0,0 +1,247 @@
+* var out = sign( 1.0, -0.9 );
+* // out => -1.0
+*/
+function sign( x, y ) {
+	return x * signum( y );
+}
```
Use `copysign` instead.

prajjwalbajpai committed:
Merge branch 'develop' into dlasv2

prajjwalbajpai committed:
chore: update the `signum` implementation and improve test coverage

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai committed:
fix: lint errors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai committed:
fix: module name in package.json

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai commented:
@kgryte I have made some changes and would like to know your thoughts on the same.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/benchmark/benchmark.js:
```diff
@@ -0,0 +1,92 @@
+*
+* @private
+*/
+function main() {
```
@prajjwalbajpai  This benchmarks needs to be refactored. The `min`/`max` iteration is pointless as the benchmark is not parameterized. We should use the unparameterized version of benchmarks.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,92 @@
+* @private
+* @returns {Function} benchmark function
+*/
+function createBenchmark( ) {
```
@prajjwalbajpai Same thing. This benchmark is incorrect.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,98 @@
+	*
+	* var out = new Float64Array( 6 );
+	*
+	* out = dlasv2( 2.0, 3.0, 4.0, out );
```
```suggestion
	* dlasv2( 2.0, 3.0, 4.0, out );
```

No need to overwrite `out` here. Applies here and throughout this PR.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,98 @@
+	* var out = new Float64Array( 6 );
+	*
+	* out = dlasv2( 2.0, 3.0, 4.0, out );
+	* // out => <Float64Array>[ 1.5513263285176897, 5.1568776039816795, 0.9664996487646696, 0.25666793515702424, 0.7496781758158659, 0.6618025632357402 ]
```
We should be using approximate values (e.g., `~1.5513`), rather than full-precision, in order to guard against implementation drift invalidating all documentation. Applies here and throughout this PR.

kgryte replied:
@prajjwalbajpai  This still needs to be updated to use approximate values. E.g.,

```js
// out => <Float64Array>[ ~1.5513, ~4.1569, ... ]
```

This comment applies here and everywhere else.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/repl.txt:
```diff
@@ -0,0 +1,98 @@
+    Parameters
+    ----------
+    F: number
+        The (0,0) element of matrix.
```
Given the listed elements, am I correct in inferring that the triangular matrix is symmetric? Otherwise, we should clarify that the matrix is a lower triangular matrix.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/repl.txt:
```diff
@@ -0,0 +1,98 @@
+    > var out =  new {{alias:@stdlib/array/float64}}( 6 );
+    > {{alias}}( 2.0, 3.0, 4.0, out );
+    > out[ 0 ]
+    1.5513263285176897
```
Same thing. Use approximate values.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/repl.txt:
```diff
@@ -0,0 +1,98 @@
+
+    While typed array views mandate a view offset based on the underlying
+    buffer, the offset parameters support indexing semantics based on starting
+    indices.
```
```suggestion
    buffer, the offset parameter support indexing semantics based on a starting
    index.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/repl.txt:
```diff
@@ -0,0 +1,98 @@
+    Returns
+    -------
+    out: Float64Array
+        Mutated array.
```
```suggestion
        Output array.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/repl.txt:
```diff
@@ -0,0 +1,98 @@
+    Returns
+    -------
+    out: Float64Array
+        Mutated array.
```
```suggestion
        Output array.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/examples/index.js:
```diff
@@ -0,0 +1,27 @@
+
+var out = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+
+out = dlasv2( 2.0, 3.0, 4.0, out );
```
This is the exact same example as used in all other documentation. Would be good to introduce a different example.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/lib/base.js:
```diff
@@ -0,0 +1,228 @@
+
+	// PMAX = 2 if G largest in absolute values
+
+	// PMAX = 3 if H largest in absolute values
```
Use a multiline comment.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/lib/base.js:
```diff
@@ -0,0 +1,228 @@
+		gasmal = true;
+		if ( ga > fa ) {
+			PMAX = 2;
+			if ( ( fa / ga ) < dlamch( 'E' ) ) {
```
`dlamch( 'E' )` is a constant. Precompute in the parent scope and use the assigned variable here.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,98 @@
+*/
+interface Routine {
+	/**
+	* Computes singular value decomposition of a 2x2 triangular matrix.
```
```suggestion
	* Computes the singular value decomposition of a 2x2 triangular matrix.
```

Applies here and throughout this PR.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/lib/ndarray.js:
```diff
@@ -0,0 +1,53 @@
+* out = dlasv2( 2.0, 3.0, 4.0, out, 1, 0 );
+* // out => <Float64Array>[ 1.5513263285176897, 5.1568776039816795, 0.9664996487646696, 0.25666793515702424, 0.7496781758158659, 0.6618025632357402 ]
+*/
+var dlasv2 = base;
```
Per conventions used elsewhere, make this a one line wrapper around `base`.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/lib/ndarray.js:
```diff
@@ -0,0 +1,53 @@
+/**
+* Computes singular value decomposition of a 2x2 triangular matrix using alternative indexing semantics.
+*
+* @type {Function}
```
```suggestion
```

After making a one line wrapper, this line can be removed.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,179 @@
+			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
+		} else {
+			delta = abs( actual[ i ] - expected[ i ] );
+			tol = rtol * EPS * abs( expected[ i ] );
```
Refactor to use `isAlmostSameValue`.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,179 @@
+	t.end();
+});
+
+tape( 'the function correctly computes svd of triangular matrix with elements in increasing order', function test( t ) {
```
The grammar in these descriptions needs to be cleaned up.


```suggestion
tape( 'the function computes the singular value decomposition of a triangular matrix with elements in increasing order', function test( t ) {
```

kgryte replied:
Similar changes apply elsewhere in this and other test files.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,179 @@
+	var out;
+
+	out = new Float64Array( 6 );
+	expected = new Float64Array( [ 1.5513263285176897, 5.1568776039816795, 0.9664996487646696, 0.25666793515702424, 0.7496781758158659, 0.6618025632357402 ] );
```
How were these values computed? Did we use a reference implementation? By hand? We should document how they were calculated. Otherwise, we run of the risk of just having used the implementation to test the implementation, which is not good.

prajjwalbajpai replied:
@kgryte I am not sure if Pranav checked these against the Netlib LAPACK code. But, I will verify it myself. Can you elaborate on how exactly do I have to document this? Do I need to attach the fortran code I used?

kgryte replied:
You can add a comment to the effect:

> Tested against the Fortran reference implementation:

And, if you want to go above and beyond, you can add a script to a `test/fixtures` folder.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,179 @@
+
+	expected = new Float64Array( [ 1.0, Infinity, 0.0, 1.0, 0.0, 1.0 ] );
+	out = new Float64Array( 6 );
+	out = dlasv2( Infinity, 2.0, 1.0, out );
```
Use PINF.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,179 @@
+	var expected;
+	var out;
+
+	expected = new Float64Array( [ 1.0, Infinity, 0.0, 1.0, 0.0, 1.0 ] );
```
Use PINF.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,179 @@
+
+	expected = new Float64Array( [ 3.0, 3.0, 0.7071067811865475, 0.7071067811865475, 0.7071067811865475, 0.7071067811865475 ] );
+	out = new Float64Array( 6 );
+	out = dlasv2( 3.0, 5e-324, 3.0, out );
```
```suggestion
	out = dlasv2( 3.0, 5.0e-324, 3.0, out );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.ndarray.js:
```diff
@@ -0,0 +1,92 @@
+	var expected;
+	var out;
+
+	expected = new Float64Array( [ 999.9, 1.2125424343860751, 999.9, 6.597707241520579, 999.9, 0.8090383403577232, 999.9, 0.5877558709457702, 999.9, 0.2452483296822559, 999.9, 0.9694602914962858 ] );
```
To improve readability, make each element its own line.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.ndarray.js:
```diff
@@ -0,0 +1,92 @@
+	t.end();
+});
+
+tape( 'the function supports complex access pattern to store computed values', function test( t ) {
```
This isn't our pattern for testing ndarray interfaces. We systematically test each parameterization in isolation. E.g., we test the base case for a stride of `1` and an offset of `0`. Then, we test a non-unit stride. Then we test a non-zero offset. Then we test negative strides. Then we test "complex access patterns". 

This is a general pattern of @Pranavchiku's PRs of taking shortcuts when testing interfaces. We will need to refactor that across his open PRs.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/README.md:
```diff
@@ -0,0 +1,224 @@
+
+# dlasv2
+
+> Compute singular value decomposition of a 2x2 triangular matrix.
```
```suggestion
> Compute the singular value decomposition of a 2x2 triangular matrix.
```

Same thing. Applies here and everywhere.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/README.md:
```diff
@@ -0,0 +1,224 @@
+var out = new Float64Array( 6 );
+
+out = dlasv2( 2.0, 3.0, 4.0, out );
+// out => <Float64Array>[ 1.5513263285176897, 5.1568776039816795, 0.9664996487646696, 0.25666793515702424, 0.7496781758158659, 0.6618025632357402 ]
```
```suggestion
var v = dlasv2( 2.0, 3.0, 4.0, out );
// returns <Float64Array>[ 1.5513263285176897, 5.1568776039816795, 0.9664996487646696, 0.25666793515702424, 0.7496781758158659, 0.6618025632357402 ]

var bool = ( v === out );
// returns true
```

If we want to communicate that the output array is returned, the return value should be assigned to a different variable name.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/README.md:
```diff
@@ -0,0 +1,224 @@
+-   **F**: the (0,0) element of matrix.
+-   **G**: the (0,1) element of matrix.
+-   **H**: the (1,1) element of matrix.
+-   **out**: [`Float64Array`][mdn-float64array] output array containing `SSMIN`, `SSMAX`, `SNR`, `CSR`, `SNL`, and `CSL` respectively.
```
```suggestion
-   **out**: [`Float64Array`][@stdlib/array/float64] output array containing `SSMIN`, `SSMAX`, `SNR`, `CSR`, `SNL`, and `CSL` respectively.
```

Prefer linking to the stdlib package.

kgryte replied:
Additionally, nowhere in the documentation do we describe what `SSMIN`, `SSMAX`, `SNR`, `CSR`, `SNL`, and `CSL` actually mean. That needs to be documented in the README, as well as the `index.d.ts` file, the `repl.txt` file, and the implementation files.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/README.md:
```diff
@@ -0,0 +1,224 @@
+
+var out = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+
+out = dlasv2( 2.0, 3.0, 4.0, out );
```
Same comment. We need a different example. This just repeats the above.

kgryte reviewed (changes requested):
This PR still needs quite a bit of clean-up before we can move it over the finish line.

prajjwalbajpai committed:
refactor: major updates to tests, benchmarks and README file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/benchmark/benchmark.js:
```diff
@@ -0,0 +1,60 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		f = uniform( -10.0, 10.0 );
```
Don't generate random variates like this within benchmarked loops; otherwise, you risk conflating the measured timings. Is the timing reflective of runtime performance? Or is it simply an artifact of generating random numbers? 

Use `random/array/uniform` outside of the loop and then do

```js
dlasv2( f[ i%f.length ], g[ i%g.length ], h[ i%h.length ], out );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,60 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		f = uniform( -10.0, 10.0 );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,60 @@
+
+// MAIN //
+
+bench( format( '%s', pkg ), function benchmark( b ) {
```
```suggestion
bench( format( '%s:ndarray', pkg ), function benchmark( b ) {
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/lib/base.js:
```diff
@@ -0,0 +1,238 @@
+		snr = srt;
+	}
+
+	// Correct signs 0f ssmax and ssmin
```
```suggestion
	// Correct signs of ssmax and ssmin
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
+		} else {
+			t.strictEqual( isAlmostSameValue( actual[ i ], expected[ i ], rtol ), true, 'returns expected value' );
+		}
```
```suggestion
		t.strictEqual( isAlmostSameValue( actual[ i ], expected[ i ], maxULPs ), true, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+* @param {Object} t - test object
+* @param {Collection} actual - actual values
+* @param {Collection} expected - expected values
+* @param {number} rtol - relative tolerance
```
```suggestion
* @param {number} maxULPs - maximum allowed ULP difference
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+* @param {Collection} expected - expected values
+* @param {number} rtol - relative tolerance
+*/
+function isApprox( t, actual, expected, rtol ) {
```
```suggestion
function isApprox( t, actual, expected, maxULPs ) {
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+
+	dlasv2( data.F, data.G, data.H, out );
+	expected = data.out;
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+
+	dlasv2( data.F, data.G, data.H, out );
+	expected = data.out;
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+
+	dlasv2( data.F, data.G, data.H, out );
+	expected = data.out;
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+
+	dlasv2( data.F, data.G, data.H, out );
+	expected = data.out;
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+
+	dlasv2( data.F, data.G, data.H, out );
+	expected = data.out;
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+
+	dlasv2( data.F, data.G, data.H, out );
+	expected = data.out;
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+	expected = new Float64Array( [ 1.0, PINF, 0.0, 1.0, 0.0, 1.0 ] );
+	out = new Float64Array( 6 );
+	out = dlasv2( PINF, 2.0, 1.0, out );
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,194 @@
+
+	dlasv2( data.F, data.G, data.H, out );
+	expected = data.out;
+	isApprox( t, out, expected, 1.0 );
```
```suggestion
	isApprox( t, out, expected, 1 );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.ndarray.js:
```diff
@@ -0,0 +1,518 @@
+* @param {Object} t - test object
+* @param {Collection} actual - actual values
+* @param {Collection} expected - expected values
+* @param {number} rtol - relative tolerance
```
All the same changes as proposed for the other test file.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.ndarray.js:
```diff
@@ -0,0 +1,518 @@
+	t.end();
+});
+
+tape( 'the function computes the singular value decomposition of a triangular matrix with elements in increasing order (Negative Stride)', function test( t ) {
```
```suggestion
tape( 'the function computes the singular value decomposition of a triangular matrix with elements in increasing order (negative stride)', function test( t ) {
```

Applies here and below.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.ndarray.js:
```diff
@@ -0,0 +1,518 @@
+	t.end();
+});
+
+tape( 'the function computes the singular value decomposition of a triangular matrix with elements in increasing order (Offset)', function test( t ) {
```
```suggestion
tape( 'the function computes the singular value decomposition of a triangular matrix with elements in increasing order (offset)', function test( t ) {
```

Applies here and below.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/README.md:
```diff
@@ -0,0 +1,262 @@
+
+<section class="intro">
+
+The singular value decomposition of a 2x2 upper triangular matrix $ \begin{bmatrix} F & G \\ 0 & H \end{bmatrix}$ is given by -
```
```suggestion
The singular value decomposition of a 2x2 upper triangular matrix $ \begin{bmatrix} F & G \\ 0 & H \end{bmatrix}$ is given by
```

kgryte reviewed (changes requested):
Left another round of comments. This is coming along.

@prajjwalbajpai Thank you for improving the tests. They are muuuuuch better. 🙌

prajjwalbajpai committed:
fix: improve benchmarks and other minor changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai commented:
Made the changes.

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.ndarray.js:
```diff
@@ -0,0 +1,518 @@
+			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
+		} else {
+			t.strictEqual( isAlmostSameValue( actual[ i ], expected[ i ], maxULPs ), true, 'returns expected value' );
+		}
```
```suggestion
		t.strictEqual( isAlmostSameValue( actual[ i ], expected[ i ], maxULPs ), true, 'returns expected value' );
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/test.dlasv2.js:
```diff
@@ -0,0 +1,190 @@
+* @param {Collection} expected - expected values
+* @param {number} maxULPs - maximum allowed ULP difference
+*/
+function isApprox( t, actual, expected, maxULPs ) {
```
I just recalled that this function is not necessary. You can replace all usage of `isApprox` below with `@stdlib/assert/is-almost-same-float64array`. Applies to this and the other test file.

prajjwalbajpai replied:
Done. One thing I found interesting was:

If I directly create the expected array as
```javascript
expected = data.out;
```
The `isAlmostSameValueFloat64Array` function was returning false. Hence, I had to wrap the array in `Float64Array`.

```javascript
expected = new Float64Array( data.out );
```

This is something that is not done in other routines where fixtures are used.

aayush0325 commented on lib/node_modules/@stdlib/lapack/base/dlasv2/examples/index.js:
```diff
@@ -0,0 +1,30 @@
+var Float64Array = require( '@stdlib/array/float64' );
+var dlasv2 = require( './../lib/' );
+
+var out = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
```
```suggestion
var out = new Float64Array( 6 );
```

aayush0325 commented on lib/node_modules/@stdlib/lapack/base/dlasv2/lib/base.js:
```diff
@@ -0,0 +1,238 @@
+	out[ offsetOut + ( 3 * strideOut ) ] = csr;
+	out[ offsetOut + ( 4 * strideOut ) ] = snl;
+	out[ offsetOut + ( 5 * strideOut ) ] = csl;
+	return out;
```
the index calculation can be done in fewer ops here

aayush0325 commented on lib/node_modules/@stdlib/lapack/base/dlasv2/README.md:
```diff
@@ -0,0 +1,262 @@
+
+<section class="intro">
+
+The singular value decomposition of a 2x2 upper triangular matrix $ \begin{bmatrix} F & G \\ 0 & H \end{bmatrix}$ is given by
```
is it just me or is the latex rendering messed up for you guys as well @kgryte @prajjwalbajpai 

<img width="1007" height="218" alt="Image" src="https://github.com/user-attachments/assets/ae42348a-8d93-4e43-b9da-3dd55f2e3664" />

prajjwalbajpai replied:
<img width="1271" height="402" alt="image" src="https://github.com/user-attachments/assets/40d7d44d-65ab-41cc-9afb-e25f1f5fbb23" />
looks fine for me

aayush0325 commented on lib/node_modules/@stdlib/lapack/base/dlasv2/test/fixtures/decreasing_order_elements.json:
it would make sense to document how you're testing using fortran in a separate repository, like [this repo](github.com/aayush0325/tests-fortran) this would help us both in terms of reviews and be a source of truth as more LAPACK PRs get queued for review over the next few weeks. imo its a good starting point, then we can consider keeping the fortran files itself in the repo but im not sure if that would achieve anything

aayush0325 reviewed (changes requested):
looks fine to me, please check the latex rendering issue!

prajjwalbajpai committed:
test: replace `isApprox` with `is-almost-same-value-float64array`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented on lib/node_modules/@stdlib/lapack/base/dlasv2/README.md:
```diff
@@ -0,0 +1,262 @@
+
+<!-- </equation> -->
+
+where `CSL`,`SNL`,`CSR`,`SNR` define orthogonal rotations and $\mathrm{SSMAX}^2, \mathrm{SSMIN}^2$ are the eigenvalues of $A^{T}A$.
```
what is `A` here? assuming its the 2x2 input matrix you can write that above

prajjwalbajpai committed:
fix: minor changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai committed:
docs: define matrix `A` clearly

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---
